### ISSUE TITLE
nullable type compiler error

### ISSUE THREAD

**Issue by emekoi** [ziglang/zig#1805]
given the following code:
```
pub const Node = struct {
    children: []?Node,
    data: ?u8,

    fn new() Node {
        return Node {
            .children = undefined,
            .data = null,
        };
    }
};
```
trying to call `Node.new()` trips [this](https://github.com/ziglang/zig/blob/master/src/analyze.cpp#L2694) assertion ~~instead of providing a meaningful error message about how the type is recursive.~~

**Comment by emekoi** [comment]
wait, isn't a slice basically a pointer with a length? so why is this type recursion?

**Comment by andrewrk** [comment]
I don't know where you're getting the idea that this is type recursion, but if you're tripping an assertion then it's a compiler bug.

**Comment by tgschultz** [comment]
The type is recursive. `children` is an *array* of Optional Nodes, each of which contains a `children` member that is an *array* of Optional Nodes, each of which contains an *array* of Optional Nodes, each of which...

Also, in the code shown `Self` is never defined. Not sure if relevant.

**Comment by emekoi** [comment]
even when i change `children` to `children: []?Node` i still get a compiler error. `children: []?*Node` works though.

**Comment by andrewrk** [comment]
That sounds like #1500

**Comment by suirad** [comment]
Unless this was cut from other code, `.data` also doesn't have a type.

**Comment by emekoi** [comment]
so should i close this then?

**Comment by andrewrk** [comment]
Proposal to fix this: #2174
